File: 0242-valid-anagram/0242-valid-anagram.py
from collections import defaultdict
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False
        
        count = [0]*26
        
        for i in xrange(len(s)):
            count[ord(s[i]) - ord('a')] += 1
            count[ord(t[i]) - ord('a')] -= 1
        
            
        for i in count:
            if i != 0:
                return False
            
        return True
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
#         if len(s) != len(t):
#             return False

#         char_map = [0] * 26

#         for i in range(len(s)):
#             char_map[ord(s[i]) - ord('a')] += 1
#             char_map[ord(t[i]) - ord('a')] -= 1

#         return all(count == 0 for count in char_map)
        

File: 0242-valid-anagram/0242-valid-anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        
        int[] charArray = new int[26];  // Array to count character frequencies
        
        for (int i = 0; i < s.length(); i++) {
            // Increase count for the character in string s
            int ind = s.charAt(i) - 'a';
            charArray[ind]++;
            
            // Decrease count for the character in string t
            ind = t.charAt(i) - 'a';
            charArray[ind]--;
        }
        
        // Check if all counts are zero
        for (int count : charArray) {
            if (count != 0) return false;
        }
        
        return true;
    }
}
